marimo/_runtime/virtual_file.py

```diff
@@ -2,49 +2,128 @@
 from __future__ import annotations
 
 import dataclasses
-import io
-from typing import TYPE_CHECKING, Callable
+import random
+import string
+import threading
+from multiprocessing import shared_memory
+from typing import TYPE_CHECKING, Optional
 
+from marimo import _loggers
 from marimo._runtime.cell_lifecycle_item import CellLifecycleItem
 
 if TYPE_CHECKING:
     from marimo._runtime.context import RuntimeContext
 
+LOGGER = _loggers.marimo_logger()
+
+
+_ALPHABET = string.ascii_letters + string.digits
+
+
+def random_filename(ext: str) -> str:
+    # adapted from: https://stackoverflow.com/questions/13484726/safe-enough-8-character-short-unique-random-string  # noqa: E501
+    # TODO(akshayka): should callers redraw if they get a collision?
+    tid = str(threading.get_native_id())
+    basename = tid + "-" + "".join(random.choices(_ALPHABET, k=8))
+    return f"{basename}.{ext}"
+
 
 @dataclasses.dataclass
 class VirtualFile:
     url: str
+    filename: str
+    buffer: bytes
+
+    def __init__(self, filename: str, buffer: bytes) -> None:
+        self.filename = filename
+        self.url = f"/@file/{filename}"
+        self.buffer = buffer
 
-    def __init__(self, filename: str) -> None:
-        raise NotImplementedError
 
+class VirtualFileLifecycleItem(CellLifecycleItem):
+    def __init__(self, ext: str, buffer: bytes) -> None:
+        self.ext = ext
+        self.buffer = buffer
+        # Not resolved until added to registry
+        self._virtual_file: Optional[VirtualFile] = None
 
-class VirutalFileLifeCycleItem(CellLifecycleItem):
-    def __init__(
-        self, filename: str, to_stream: Callable[[], io.BytesIO]
-    ) -> None:
-        self.filename = filename
-        self.to_stream = to_stream
+    @property
+    def virtual_file(self) -> VirtualFile:
+        assert self._virtual_file is not None
+        return self._virtual_file
 
     def create(self, context: "RuntimeContext") -> None:
-        context.virtual_file_registry.add(
-            self.filename, self.to_virtual_file()
-        )
+        filename = random_filename(self.ext)
+        registry = context.virtual_file_registry
+        # create a unique filename for the virtual file
+        tries = 0
+        max_tries = 100
+        while registry.has(filename) and tries < max_tries:
+            filename = random_filename(self.ext)
+            tries += 1
+        if tries > max_tries:
+            raise RuntimeError(
+                "Failed to add virtual file to registry. "
+                "This is a bug in marimo. Please file an issue."
+            )
```

Comment:
@mscolnick I decided to force virtual file names to be unique, since we don't have a use-case for deterministic names yet. Let me know if this is ok?

Reply:
Yea that's totally fine. I agree we don't need to pre optimize that